File: 2023/day13/ruby/solution.rb
#!/usr/bin/env ruby

def parse_input(text)
  text.strip.split("\n\n").map { |block| block.split("\n") }
end

def count_differences(str1, str2)
  str1.chars.zip(str2.chars).count { |c1, c2| c1 != c2 }
end

def find_vertical_reflection(pattern, target_diff:)
  return 0 if pattern.empty?

  width = pattern.first.length

  (1...width).find do |col|
    total_diff = pattern.sum do |row|
      left = row[0...col].reverse
      right = row[col..]
      min_len = [left.length, right.length].min
      count_differences(left[0...min_len], right[0...min_len])
    end
    total_diff == target_diff
  end || 0
end

def find_horizontal_reflection(pattern, target_diff:)
  return 0 if pattern.empty?

  (1...pattern.length).find do |row|
    top = pattern[0...row].reverse
    bottom = pattern[row..]
    compare_len = [top.length, bottom.length].min

    total_diff = top.take(compare_len).zip(bottom.take(compare_len)).sum do |top_row, bottom_row|
      count_differences(top_row, bottom_row)
    end

    total_diff == target_diff
  end || 0
end

def summarize_pattern(pattern, target_diff:)
  vertical = find_vertical_reflection(pattern, target_diff: target_diff)
  return vertical if vertical > 0

  horizontal = find_horizontal_reflection(pattern, target_diff: target_diff)
  horizontal * 100
end

def part1(patterns)
  patterns.sum { |pattern| summarize_pattern(pattern, target_diff: 0) }
end

def part2(patterns)
  patterns.sum { |pattern| summarize_pattern(pattern, target_diff: 1) }
end

def main
  input_file = File.join(__dir__, '..', 'input.txt')
  text = File.read(input_file)
  patterns = parse_input(text)

  puts "Part 1: #{part1(patterns)}"
  puts "Part 2: #{part2(patterns)}"
end

main if __FILE__ == $PROGRAM_NAME

File: 2023/day13/python/solution.py
#!/usr/bin/env python3
import sys
from pathlib import Path

def parse_input(text):
    """Parse input into list of patterns (each pattern is a list of strings)."""
    blocks = text.strip().split('\n\n')
    return [block.split('\n') for block in blocks]

def find_vertical_reflection(pattern):
    """Find vertical line of reflection. Returns columns to the left, or 0 if none."""
    if not pattern:
        return 0
    width = len(pattern[0])

    for col in range(1, width):
        # Check if reflection exists at this column
        is_reflection = True
        for row in pattern:
            # Compare left side with right side (mirrored)
            left = row[:col][::-1]  # reverse left side
            right = row[col:]
            # Compare the overlapping parts
            min_len = min(len(left), len(right))
            if left[:min_len] != right[:min_len]:
                is_reflection = False
                break
        if is_reflection:
            return col
    return 0

def find_horizontal_reflection(pattern):
    """Find horizontal line of reflection. Returns rows above, or 0 if none."""
    if not pattern:
        return 0
    height = len(pattern)

    for row in range(1, height):
        # Check if reflection exists at this row
        is_reflection = True
        # Compare top with bottom (mirrored)
        top = pattern[:row][::-1]  # reverse top side
        bottom = pattern[row:]
        min_len = min(len(top), len(bottom))
        for i in range(min_len):
            if top[i] != bottom[i]:
                is_reflection = False
                break
        if is_reflection:
            return row
    return 0

def summarize_pattern(pattern):
    """Get the summary value for a pattern."""
    v = find_vertical_reflection(pattern)
    if v > 0:
        return v
    h = find_horizontal_reflection(pattern)
    return h * 100

def part1(patterns):
    """Calculate the sum of all pattern summaries."""
    return sum(summarize_pattern(p) for p in patterns)

def count_differences(s1, s2):
    """Count character differences between two strings."""
    return sum(c1 != c2 for c1, c2 in zip(s1, s2))

def find_vertical_reflection_with_smudge(pattern):
    """Find vertical line with exactly one smudge fix needed."""
    if not pattern:
        return 0
    width = len(pattern[0])

    for col in range(1, width):
        total_diff = 0
        for row in pattern:
            left = row[:col][::-1]
            right = row[col:]
            min_len = min(len(left), len(right))
            total_diff += count_differences(left[:min_len], right[:min_len])
            if total_diff > 1:
                break
        if total_diff == 1:
            return col
    return 0

def find_horizontal_reflection_with_smudge(pattern):
    """Find horizontal line with exactly one smudge fix needed."""
    if not pattern:
        return 0
    height = len(pattern)

    for row in range(1, height):
        total_diff = 0
        top = pattern[:row][::-1]
        bottom = pattern[row:]
        min_len = min(len(top), len(bottom))
        for i in range(min_len):
            total_diff += count_differences(top[i], bottom[i])
            if total_diff > 1:
                break
        if total_diff == 1:
            return row
    return 0

def summarize_pattern_with_smudge(pattern):
    """Get the summary value for a pattern with smudge fix."""
    v = find_vertical_reflection_with_smudge(pattern)
    if v > 0:
        return v
    h = find_horizontal_reflection_with_smudge(pattern)
    return h * 100

def part2(patterns):
    """Calculate the sum with smudge fixes."""
    return sum(summarize_pattern_with_smudge(p) for p in patterns)

def main():
    input_file = Path(__file__).parent.parent / 'input.txt'
    text = input_file.read_text()
    patterns = parse_input(text)

    print(f"Part 1: {part1(patterns)}")
    print(f"Part 2: {part2(patterns)}")

if __name__ == '__main__':
    main()
